### ISSUE TITLE
Query returns inconsistent results

### ISSUE THREAD

**Issue by wdroste** [duckdb/duckdb#12057]
### What happens?

If we run this query several times

### To Reproduce

The tables are structured to support a graph. There's are tables with data and then one table w/ relationships to map the between objects in x table.

Vulnerability - [ targets ] -> Host 

Where Vulnerability and Host are tables and there's a relationship called `targets` that links the two.

This is the query below that includes `OR` clauses to include users that filter the data.

The CLI comes back with different results if you run it a few times. We randomly get different results it maybe 370 but then it comes w/ 371 or 369 as a result.

```
with "cte" as (select v.id                "v_id",
                      list(distinct h.id) FILTER(WHERE h.id IS NOT NULL) "h_id",
                       list(distinct i.id) FILTER(WHERE i.id IS NOT NULL) "i_id",
                       any_value(r.id) "r_id",
                      list(distinct u.id) FILTER(WHERE u.id IS NOT NULL) "u_id",
                       list(distinct u2.id) FILTER(WHERE u2.id IS NOT NULL) "u2_id",
                       list(distinct h.name) FILTER(WHERE h.name IS NOT NULL) "h_name",
                       v.displayName "v_displayName"
               from "Vulnerability" "v"
                        left outer join "relationships" "_relationships_1_" on (v.id = _relationships_1_.startNode and _relationships_1_.associationName = 'informedUsers')
                        left outer join "InformedUser" "i" on _relationships_1_.endNode = i.id
    left outer join "relationships" "_relationships_2_" on (i.id = _relationships_2_.startNode and _relationships_2_.associationName = 'members')
    left outer join "User" "u2" on _relationships_2_.endNode = u2.id
    left outer join "relationships" "_relationships_3_" on (v.id = _relationships_3_.startNode and _relationships_3_.associationName = 'targets')
    left outer join "Host" "h" on _relationships_3_.endNode = h.id
    left outer join "relationships" "_relationships_4_" on (v.id = _relationships_4_.startNode and _relationships_4_.associationName = 'remediationOwner')
    left outer join "RemediationOwner" "r" on _relationships_4_.endNode = r.id
    left outer join "relationships" "_relationships_5_" on (r.id = _relationships_5_.startNode and _relationships_5_.associationName = 'members')
    left outer join "User" "u" on _relationships_5_.endNode = u.id
where true
group by v.id, v.displayName
    )
select count (distinct "h_id") "Count unique (Targets)"
from (
         select
             unnest("h_id") "h_id",
             unnest(list_filter(u_id, __element__ -> "__element__" = 1788948092043718656)) "u_id",
             unnest(list_filter(u2_id, __element__ -> "__element__" = 1788948092043718656)) "u2_id",
             unnest(list_filter(h_name, __element__ -> "__element__" like ('%brinqa%'))) "h_name",
             "v_displayName"
         from "cte"
     ) "alias_79071577"
where ("h_id" is not null
    and ("u_id" = 1788948092043718656
        or "u2_id" = 1788948092043718656)
    and ("v_displayName" like ('%UDP%')
        or "h_name" like ('%brinqa%')))
limit 3
offset 0;
```

### OS:

Linux

### DuckDB Version:

10.1, 10.2

### DuckDB Client:

CLI

### Full Name:

Will Droste

### Affiliation:

Brinqa

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a stable release

### Did you include all relevant data sets for reproducing the issue?

No - I cannot share the data sets because they are confidential

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by wdroste** [comment]
We've kind of narrowed it down to the use of distinct.

**Comment by soerenwolfers** [comment]
Can you replace all confidential data with random data of the same data type? I imagine not much can be done for you based on just a massively complex query and not even a description of the schemata and sizes of your inputs.

**Comment by szarnyasg** [comment]
Hi @wdroste, thanks for the report. Indeed as @soerenwolfers remarked, we cannot move forward without a reproducible example.

**Comment by wdroste** [comment]
Thanks for taking a look at this so quickly. I'm working w/ engineering to get the a suitable database for you to reproduce the problem. I realize its a complex query, so again I appreciate you even looking at it.

**Comment by hannes** [comment]
Thanks for providing the database file, I've tried running your query, but it seems to consistently return 0. Can you share a bit more info about your environment please?

**Comment by wdroste** [comment]
My apologies, here's the database that fits that query it was a missing some data that results in a zero outcome as you've seen. With this database the expected outcome is 370.. however it fluctuates between 370 and 371 with enough invocations. 

https://github.com/brinqa/test-duckdb/raw/1.x/testForACLs.db.7z

**Comment by hannes** [comment]
Indeed I see it fluctuating too on OSX with 0.10.2 now, also with latest `main`.

**Comment by hannes** [comment]
So after much head scratching this is not actually a bug but rather expected behaviour. I'm going to use a small example here to explain:

Let's create a table with two columns of random numbers, 1000 rows:
```SQL
SELECT setseed(0);
CREATE OR REPLACE TEMPORARY TABLE numbers AS 
  SELECT 
   (random()*100)::INTEGER r1, 
   (random()*100)::INTEGER r2 
  FROM range(1000);
```

This table looks something like this:
| r1 | r2 |
|---:|---:|
| 29 | 52 |
| 37 | 89 |
| 57 | 24 |
| 58 | 80 |
| 42 | 35 |

Let's run a query that checks a filter *across multiple columns*

```SQL
SELECT COUNT(DISTINCT *) FROM numbers WHERE r1 > 90 AND r2 > 90;
```

This returns 13 for me, consistently.

Now let's simulate your query:

```SQL
WITH cte AS (
  SELECT 
    LIST(DISTINCT r1) l1, 
    LIST(DISTINCT r2) l2 
  FROM numbers)
SELECT COUNT(*) FROM (
  SELECT 
    UNNEST(l1) r1, 
    UNNEST(l2) r2 FROM cte ) 
  WHERE r1 > 90 AND r2 > 90;
```

when I re-run this a couple of times, I get either 0, 1, 2, or 3 as result randomly. How can that be? The reason for this is the parallel processing of this query. To check `DISTINCT`ness, we use a hash table in separate threads, which is then merged together. But the order in which those are merged is not defined, which means the *order of list elements in `l1` and `l2` is now undefined*. 

When we later use `UNNEST` to turn the lists back into rows, we have lost the original relation between values in rows, and recombine them rather randomly.

We can see this also by just looking at the result of the CTE:

```SQL
WITH cte AS (
SELECT 
  LIST(DISTINCT r1) l1, 
  LIST(DISTINCT r2) l2 
FROM numbers)
SELECT l1[0:5], l2[0:5] FROM cte;
```

If I re-run this a couple of times, I get

|       l1[0:5]        |       l2[0:5]       |
|----------------------|---------------------|
| [37, 62, 64, 79, 54] | [49, 2, 39, 34, 10] |

and

|       l1[0:5]       |       l2[0:5]        |
|---------------------|----------------------|
| [22, 0, 83, 81, 16] | [52, 81, 83, 31, 32] |

and 

|       l1[0:5]        |       l2[0:5]        |
|----------------------|----------------------|
| [29, 34, 10, 93, 71] | [50, 42, 74, 70, 25] | 

for example.

The issue predictably disappears by disabling parallelism by setting `SET threads=1`. Then, the query consistently returns 1.

In conclusion, the semantics of this query probably have to be re-worked to not lose the connections between rows if the filter is to be applied after `UNNEST`.

**Comment by wdroste** [comment]
Thank you, this was very helpful.
